### ISSUE TITLE
Resolution of native libraries when using DllImport

### ISSUE THREAD

**Issue by antoinebj** [dotnet/runtime#54645]
TL;DR: How can I make `DllImport` automatically load the right native library from the `runtimes/<rid>/native` path when it is in a project referred by `ProjectReference`, similar to what `PackageReference` enables by adding the appropriate resolution information in deps.json?

--

If I create a NuGet package and put native DLLs in `runtimes/<rid>/native/`, the deps.json file generated for the assembly referencing that package will get `runtimeTargets` entries for the native libraries, allowing `DllImport` to find the appropriate DLLs to load. That mechanism is very useful for easily targeting multiple platforms with "Any CPU" despite reliance on native libraries that are each specific to a platform/architecture combination.

Now I would like to unit-test the classes in that package. Usually I put the unit test project in the same solution, make it refer to the project containing the classes to test. Unfortunately, that doesn't bring the mechanisms from `PackageReference`, as it only uses a `ProjectReference`. Therefore the deps.json for the unit test assembly doesn't get hints for the native libraries, and when the unit test is run, it complains about not finding the native DLLs.

A workaround would be to have a separate NuGet package containg nothing more than the native libraries, making the project containing .NET code refer to it, similar to what libgit2sharp does. However that's not convenient when updating the native libraries because it requires publishing their package before being able to test their integration in the .NET code.
I know about the NativeLibrary class where one can register a custom resolver, but that seems tedious for true multiplatform. I didn't find a simple way to replicate the RID-based resolution.

Is there any other way? Any thoughts?

**Comment by vitek-karas** [comment]
@marcpopMSFT I think this is very much an SDK issue. 
* SDK needs to communicate information about RID specific assets to the runtime. 
* Such communication channel already exists, it's `.deps.json` (it's basically the main reason why this file exists today)
* SDK will correctly populate `.deps.json` if the RID specific assets are coming from NuGet packages
* SDK doesn't have mechanisms to define RID specific assets in non-NuGet sources (the project itself, dependent P2P references, ...) - and inherently it doesn't have a mechanism to correctly populate this information into `.deps.json`

**Comment by alexrp** [comment]
> * SDK doesn't have mechanisms to define RID specific assets in non-NuGet sources (the project itself, dependent P2P references, ...) - and inherently it doesn't have a mechanism to correctly populate this information into `.deps.json`

FWIW, having a way to do this is essential for parts of https://github.com/alexrp/zig-msbuild-sdk/issues/8 as well.

**Comment by marcpopMSFT** [comment]
@vitek-karas Isn't this an issue in deps.json creation which from our discussion, the runtime team controls (hence our decision to move it to runtime repo)?

**Comment by jkoritzinsky** [comment]
@marcpopMSFT I agree with @vitek-karas that this is a dotnet/sdk issue.

The .deps.json format already has a representation for the required concept.

The task that generates the .deps.json is in the dotnet/sdk repo.

The .NET SDK needs to provide a (either add or document an existing) mechanism for developers to specify native assets for a given RID. This work is definitely SDK work as the experience is in the SDK itself, not in the runtime.

I'll let you two determine where this ends up. In any case, I think this falls into either Loader or SDK since those are the areas that would own the discovery mechanism, not Interop.

**Comment by vitek-karas** [comment]
While I agree that the final piece of code which writes this info into `.deps.json` can be owned by the runtime (although in the last several years we were not the owners of this), I think the primary problem is in defining the desired behavior. Which is in my mind solely an SDK thing. The SDK needs to define what is the right way to declare RID specific assets in cases others then `PackageReference` - this doesn't exist anywhere and it's not just about `.deps.json` writing, this information also needs to be correctly handled by things like RID targeted build/publish (when the SDK produces RID specific application, it should only copy the RID specific assets which are for that RID).

**Comment by marcpopMSFT** [comment]
@richlander has been digging into the RID behaviors and so may have an opinion here on how this should work long term.

**Comment by antoinebj** [comment]
Since this has been marked as a feature request, could you please advise on a better title?
"RID-based native library resolution for DllImport when using ProjectReference"?
If you also want me to put a clearer intro to the issue, feel free to advise.

**Comment by vitek-karas** [comment]
@antoinebj Thanks a lot for reminding us about this.

I think the core of the problem is lack of support for RID specific assets our project system. That is things like:
* How do I define a RID specific asset in a classlib/app (not NuGet package)
* ProjectReference should be able to consume classlib with RID specific assets correctly
* RID specific assets from all sources (app project itself, ProjectReferences and PackageReferences) are correctly copied to the output (for portable apps into runtime subdirectory, for RID specific apps into the root) and the correct `.deps.json` is produced.

I would suggest renaming this issue to something like "Support RID specific assets outside of NuGet packages" (or something similar).

I'll move this issue back to the SDK repo where it belongs.
